spec.bs

```diff
@@ -2664,6 +2691,81 @@ To <dfn>fetch trusted signals</dfn> given a [=URL=] |url|, an [=origin=] |script
   1. Return « |signals|, |perInterestGroupData|, |dataVersion| ».
 </div>
 
+<div algorithm>
+To <dfn>fetch trusted key value signals</dfn> given a [=URL=] |url|, a [=byte sequence=] |body|, a string
+|context|, an [=origin=] |scriptOrigin|, a [=policy container=] |policyContainer|, an |indexMap|, and a
+[=boolean=] |isBiddingSignal|:
+
+1. Let |request| be a new [=request=] with the following properties:
+  :   [=request/URL=]
+  ::  |url|
+  :   [=request/method=]
+  ::  `POST`
+  :   [=request/body=]
+  ::  |body|
+  :   [=request/origin=]
+  ::  |scriptOrigin|
+  :   [=request/header list=]
+  ::  «`Accept`: `application/json`»
+  :   [=request/client=]
+  ::  `null`
+  :   [=request/mode=]
+  ::  "`cors`"
+  :   [=request/referrer=]
+  :: "`no-referrer`"
+  :   [=request/credentials mode=]
+  ::  "`omit`"
+  :   [=request/redirect mode=]
+  :: "`error`"
+  :   [=request/policy container=]
+  ::  A new [=policy container=] whose [=policy container/IP address space=] is |policyContainer|'s
+    [=policy container/IP address space=]
+
+  Issue: One of the side-effects of a `null` client for this subresource request is it neuters all
+  service worker interceptions, despite not having to set the service workers mode.
+
+1. Let |resultList| be the result of deserializing |responseBody| using |context|. The
+  deserialization method may follow that described in
+  [Section 2.3.6 of the Protected Audience Key Value Services](https://privacysandbox.github.io/draft-ietf-protected-audience-key-value-service/draft-ietf-protected-audience-key-value-services.html#name-parsing-a-response).
+1. Let |signals| be null.
+1. Let |dataVersion| be an empty [=ordered map=].
+1. Let |perInterestGroupData| be an [=ordered map=].
+1. [=Fetch=] |request| with [=fetch/useParallelQueue=] set to true, and
+  [=fetch/processResponseConsumeBody=] set to the following steps given a [=response=] |response|
+  and null, failure, or a [=byte sequence=] |responseBody|:
+  1. If [=validate fetching response=] with |response|, |responseBody| and
+    "`message/ad-auction-trusted-signals-response`" returns false, set |signals| to failure and return.
+  1. [=list/For each=] |result| in |resultList|:
+    1. If |isBiddingSignal| is true:
+      1. [=map/For each=] |name| → |value| in |result|["interestGroupNames"]:
+        1. If |indexMap|[|name|] does not equal to |result|["index"], set |signals| to failure and return.
+        1. Set |perInterestGroupData|[|name|] to |value|.
+        1. If |result| [=map/contain=] "dataVersion":
+          1. If |result|["dataVersion"] is not an integer, or is less than 0 or more than
+            2<sup>32</sup>&minus;1, set |signals| to failure and return.
+          1. Otherwise, set |dataVersion|[|name|] to |result|["dataVersion"].
+      1. [=map/For each=] |key| → |value| in |result|["keys"]:
+        1. Set |siganls|[|key|] to |value|.
+    1. Otherwise:
+      1. [=map/For each=] |url| → |value| in |result|["renderUrls"]:

```

Comment:
are we expecting strings or [=URLs=] as |url|'s data type?
From later usage of the returned |dataVersion|, it seems it expects an URL

Reply:
Yes, the |url| here is type of [=URL=]. Do we need to specify the type when for each loop the map with `key` → `value` in `map`?

Reply:
we don't need to. But the key |url| you put into |signals| is passed to all trusted bidding signals (see step 4.1 of https://wicg.github.io/turtledove/#fetch-the-current-outstanding-trusted-signals-batch), whose key is string, not url. You need to make their key match at some point.

Reply:
The |url| is certain as `string`. Added `[=URL serializer|serialized=]` in:
```
1. Set |result|'s [=trusted scoring signals reply/data version=] to |scoringDataVersion|
          [|entry|'s [=URL serializer|serialized=] [=trusted scoring signals request/render URL=]].
```
in trusted scoring signals usage.